scripts/systemd/check_and_upgrade_serial_keel.sh

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+# This script will either setup a fresh serial-keel service instance, or upgrade your existing one.
+# This will also be executed during every bootup, so a check for an upgrade will be performed.
+set -e
+
+mkdir -p $HOME/.config/systemd/user/
+mkdir -p $HOME/.config/environment.d/
+
+# The name of the branch can be specified as a argument, or it defaults to 'main'
+BRANCH=${1-main}
+CONFIG_PATH=${2-$HOME/config.ron}
+ENV_FILE=$HOME/.config/environment.d/serial-keel.conf
+
+if ! test -f $ENV_FILE; then
+    env > $ENV_FILE
+fi
+
+SCRIPT_PATH="$( cd -- "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
+cd $SCRIPT_PATH
+
+REPO_ROOT=$(git rev-parse --show-toplevel)
+cd $REPO_ROOT
+
+# Check if there are new changes. If there are none, and if the serial-keel service
+# already exists on the machine, then start the serial-keel server
+
+GIT_NO_BRANCH_SWITCH=true
+if [[ $(git rev-parse --abbrev-ref HEAD) != "$BRANCH" ]]; then
+    GIT_NO_BRANCH_SWITCH=false
+    git checkout $BRANCH
+fi
+
+git fetch
+
+GIT_NO_NEW_REMOTE_CHANGES=true
+if git status -uno | grep -q "Your branch is behind"; then
+    GIT_NO_NEW_REMOTE_CHANGES=false
+    git pull
+fi
+
+
+if $GIT_NO_NEW_REMOTE_CHANGES && $GIT_NO_BRANCH_SWITCH && test -f $HOME/.config/systemd/user/serial-keel.service; then
+    echo "serial-keel already up to date."
+    exit 0
+fi
+
+echo "We have a serial-keel update to perform.."
+
+echo "STEP 1/4: Copying the serial-keel.service file, \
+in case we have a fresh installation, or if there were updates"
+
+# This replaces all occurences of [user-name-here] and [branch-here] in the serial-keel.template file and 
+# with the current user and branch name and places the service file in /etc/systemd/system
+sed "s|ExecStartPre=.*|ExecStartPre=$REPO_ROOT\/scripts\/systemd\/check_and_upgrade_serial_keel.sh $BRANCH|;
+     s|ExecStart=.*|ExecStart=$HOME\/.cargo\/bin\/serial-keel $CONFIG_PATH|" $REPO_ROOT/scripts/systemd/serial-keel.template.service > $HOME/.config/systemd/user/serial-keel.service
+
+systemctl --user enable serial-keel.service
+systemctl --user daemon-reload
+
+echo "STEP 2/4: Stopping serial-keel server if already running.."
+systemctl --user stop serial-keel || true
```

Comment:
So I did some testing and I think we have to split things up a bit.

1. A single "one time script" to do the copying of the template service file and enabling the service
2. Everything else: Check git, upgrade if needed, start the server

Because I think that doing e.g. `systemctl --user stop serial-keel` as part of the `serial-keel` systemd service itself is a strange "self reference" which kills its own upgrade.

**How to see this behaviour:**
First I ran the script to install the service.

Notice I pushed an empty commit to your branch.
This allows me locally then `git reset --hard` to the commit _before_ the empty commit.
If I then do `systemctl --user restart serial-keel` it then triggers the update because it sees the upstream has a never revision (since `reset --hard` removes the local one).

It will then run successfully until after
`echo "STEP 2/4: Stopping serial-keel server if already running.."`
Then the update process gets killed, and the status (`systemctl --user status serial-keel`) says:

```text
Feb 25 11:36:19 pop-os check_and_upgrade_serial_keel.sh[6474]: STEP 2/4: Stopping serial-keel server if already running..
Feb 25 11:36:19 pop-os systemd[2677]: serial-keel.service: Control process exited, code=killed, status=15/TERM
Feb 25 11:36:19 pop-os systemd[2677]: serial-keel.service: Failed with result 'signal'.
Feb 25 11:36:19 pop-os systemd[2677]: Stopped Serial Keel Server.
```

Which leads me to think `systemctl --user stop serial-keel || true` sent a terminate signal to _itself_.

Therefore I think we need a cleaner separation where the `serial-keel` script has no references to `systemctl`.

Similarly, it shouldn't make sense to finish off the script with `systemctl --user start serial-keel`, since if we are assuming that it's `systemd` that is running this script then it should already be in the context of starting (or restarting) the service.

Let me know if you have any questions.

Reply:
wouldn't it be better to skip step number 2 and on step 4, instead just do `systemctl --user restart serial-keel`? I assume that would achieve the same result and you can still keep it in one script?

Reply:
I suppose then, we will have an issue with overwriting the serial-keel binary when it is running. I think I'll split the script into two, one which is only triggered by the user - and the other run as a pre-execution script by the service.

Reply:
@Mangl10 I think there shouldn't be any calls to `systemctl` except the first time when the service is installed.
When the script is called after the first time we assume it is being called by `systemd`, which handles starting/restarting by itself, so we don't have to.

Therefore I think a valid strategy is to either:

1. Keep everything in one script, but have a guard which checks if this is the first time we are running the script (e.g. by looking for the service file), and only do `systemctl` calls within this guard, or
2. Split it up into two scripts, one for service setup, and one for running the upgrade+server itself.

@stephen-nordic I don't that will be an issue.
Remember, when `systemd` is (re)starting the service it will stop it first (if it was active) which would first stop the server.
That is not the same as starting the service and then running the script locally- that isn't a use case we need or want to support.

Reply:
@torsteingrindvik Ah yes, I see that makes sense. So @Mangl10's suggestion would work.

But if we were to split the script into two, (or within the guard for first-time script run in the single script method) I don't think we would be able to eliminate all calls to systemctl in the upgrade script. What if there was a change in the .service file itself? In that case, we would need to update the .service file, do a systemctl daemon-reload and restart the service, both of which would involve calls to systemctl.

Or do we not want to support automatic .service file updates entirely? We could make it such that .service file changes can only be carried out by manually.

I'd like to propose this:
 1. Option 1 - Keeping everything in one script, and have a check to see if we are running the script for the first time. We will have systemctl calls only within this check. If the .service file has changed, the user has to run this script manually passing a special flag, which will run the script as if it were a first time installation, stop the serial-keel service and replace the service file.

Reply:
Right now I'm thinking it's fine to just update the service manually, that shouldn't happen often.

So in script 1 just copy the template and overwrite if the service exists etc.
And in script 2 perhaps call script 1 if first time, and else skip that part.

I don't think automating the service upgrade itself is worth it for now, our main goal is just to make sure the SK server is up to date.

Reply:
Thanks @torsteingrindvik and @Mangl10 :slightly_smiling_face: I've added a commit that should work. Let me know  if there are any more concerns. :+1: